Add detect_from_csvs and detect_from_dataframes methods to MultiTableMetadata

R-Palazzo opened the pull request:
Resolve #1520

fealho commented on sdv/metadata/multi_table.py:
```diff
@@ -343,6 +346,20 @@ def detect_table_from_dataframe(self, table_name, data):
+                the values are the dataframes.
+        """
+        if not data or not all(isinstance(df, pd.DataFrame) for df in data.values()):
+            raise ValueError('The provided dictionary must contain only pandas DataFrame objects')
```
End with a period.

fealho commented on sdv/metadata/multi_table.py:
```diff
@@ -343,6 +346,20 @@ def detect_table_from_dataframe(self, table_name, data):
+
+        Args:
+            data (dict):
+                Dictionary of ``pandas.DataFrame`` objects where the keys are the table names and
```
You can simplify to `Dictionary of table names to dataframes.

fealho commented on sdv/metadata/multi_table.py:
```diff
@@ -354,11 +371,30 @@ def detect_table_from_csv(self, table_name, filepath):
+            folder_name (str):
+                Name of the folder to detect the metadata from.
+
+        Raises:
```
We never write `raises` in docstrings. I guess for consistency it would make sense not to add it?

fealho commented on sdv/metadata/multi_table.py:
```diff
@@ -354,11 +371,30 @@ def detect_table_from_csv(self, table_name, filepath):
+        Raises:
+            ValueError: If no CSV files are detected in the folder.
+        """
+        csv_files = [filename for filename in os.listdir(folder_name) if filename.endswith('.csv')]
```
You could validate the folder name exists if you want, up to you.

fealho commented on sdv/metadata/multi_table.py:
```diff
@@ -354,11 +371,30 @@ def detect_table_from_csv(self, table_name, filepath):
+            raise ValueError(f"No CSV files detected in the folder '{folder_name}'")
+
+        for filename in csv_files:
+            table_name = filename[:-4]  # Removing the .csv extension
```
Could do this directly in line 388.

R-Palazzo replied:
I think this is fine here because I need the `filename` and the `table_name` after

fealho commented on sdv/metadata/multi_table.py:
```diff
@@ -354,11 +371,30 @@ def detect_table_from_csv(self, table_name, filepath):
+        csv_files = [filename for filename in os.listdir(folder_name) if filename.endswith('.csv')]
+
+        if not csv_files:
+            raise ValueError(f"No CSV files detected in the folder '{folder_name}'")
```
End with period, pretty much all our error messages do, so it makes sense to be consistent.

fealho commented on tests/unit/metadata/test_multi_table.py:
```diff
@@ -1495,6 +1494,54 @@ def test_detect_table_from_csv_table_already_exists(self):
+
+        # Run and Assert
+        expected_message = re.escape("No CSV files detected in the folder '{}'".format(tmp_path))
+
```
New line not necessary.

fealho commented on tests/integration/metadata/test_multi_table.py:
```diff
@@ -132,3 +135,103 @@ def test_upgrade_metadata(tmp_path):
+                }
+            }
+        },
+        'relationships': [],
```
We will add the logic to detect the relationships in another issue, correct?

R-Palazzo replied:
Yes

fealho reviewed (commented):
Looks great!

R-Palazzo commented:
Thanks for your review @fealho. I addressed the comments in  [b85bf4f](https://github.com/sdv-dev/SDV/pull/1533/commits/b85bf4f192a8041d51ff077291aaa9eea896db42).

pvk-developer commented on sdv/metadata/multi_table.py:
```diff
@@ -354,11 +370,33 @@ def detect_table_from_csv(self, table_name, filepath):
+                Name of the folder to detect the metadata from.
+
+        """
+        if os.path.exists(folder_name) and os.path.isdir(folder_name):
+            csv_files = [
+                filename for filename in os.listdir(folder_name) if filename.endswith('.csv')
+            ]
```
Instead of using `os.path` we can use `Pathlib` which is more powerful and will improve the code down the line.
For example:
```python
Path().rglob(f'{folder_name}/*.csv')
```
will return a list with all the paths of csv files-

R-Palazzo replied:
This is really good advice, thanks! Done in [4c37e19](https://github.com/sdv-dev/SDV/pull/1533/commits/4c37e19e8e25d340c6c5693b237d1e46fe0d9c28).

pvk-developer commented on sdv/metadata/multi_table.py:
```diff
@@ -354,11 +370,33 @@ def detect_table_from_csv(self, table_name, filepath):
+
+        for filename in csv_files:
+            table_name = filename[:-4]  # Removing the .csv extension
+            csv_file = os.path.join(folder_name, filename)
```
Wouldn't need this since `pathlib` will have the full path already.

fealho reviewed (approved):
👍

pvk-developer reviewed (approved):
👍🏻 LGTM!

R-Palazzo commented on sdv/metadata/multi_table.py:
```diff
@@ -354,11 +370,32 @@ def detect_table_from_csv(self, table_name, filepath):
         table = SingleTableMetadata()
-        data = table._load_data_from_csv(filepath)
-        table._detect_columns(data)
+        table.detect_from_csv(filepath)
```
I think we need this change because `_load_data_from_csv` no longer exists in `SingleTableMetadata`. I added an integration test in [922aad0](https://github.com/sdv-dev/SDV/pull/1533/commits/922aad0898d17abab81f7b05f0231e5647658972) for the `detect_table_from_csv` which fails in the master branch

amontanez24 replied:
It got moved here
https://github.com/sdv-dev/SDV/blob/387e9dd949abc0c412bd6e4b65e489cee1c93c08/sdv/utils.py#L168
I think we don't want to call `detect_from_csv` because if the error gets raised it says
https://github.com/sdv-dev/SDV/blob/387e9dd949abc0c412bd6e4b65e489cee1c93c08/sdv/metadata/single_table.py#L287-L289

R-Palazzo replied:
I see, thanks for the explanation, done in [ef79cf8](https://github.com/sdv-dev/SDV/pull/1533/commits/ef79cf86efbe222c76cd3b9c397240b0b94180cd)

R-Palazzo commented on tests/integration/metadata/test_multi_table.py:
```diff
@@ -132,3 +135,103 @@ def test_upgrade_metadata(tmp_path):
+
+    metadata = MultiTableMetadata()
+
+    with tempfile.TemporaryDirectory() as temp_dir:
```
Yes, done in [8cf7f02](https://github.com/sdv-dev/SDV/pull/1533/commits/8cf7f02112b37da42eb5fc355dca1311d25965a3)

amontanez24 reviewed (approved):
Thanks for addressing the comments! 🚢 📦
